Fix IPv6-only DNS by checking IPv6 first if have a public scope address

sgryphon opened the pull request:
## Description of Change

On top of the IPv6 change to use `getaddrinfo()`, this implements a fix for IPv6-only networks by checking if we have an IPv6 public scope address, and if we do then check IPv6 first.  If there is no IPv6-address, or we are on an IPv4-only network, then the change has no effect.

By checking IPv6 first when available, it allows devices to work across all network types -- IPv4-only, IPv6-only, and dual stack, connecting to all destination types -- IPv4-only, IPv6-only, and dual-stack servers.

Previously, while IPv6-only to IPv6-only worked, trying to connect from IPv6-only to dual stack (including DNS64) did not work, as the IPv4 address was always returned (and unreachable).

Note that this change does not do much unless IPv6 is enabled (Libs change): it will simply change the dual-stack preference order to IPv6-first, but IPv6-only will still fail if there is no DNS to query.

## Tests scenarios

I have tested on a M5Stack Core2, with an ESP32 chip.

I have tested across different network types, and different destination types. Now that the TLS IPv6 support bug is merged, this now works across both HTTP and HTTPS connections.

Note: If IPv6 is never enabled, then the code is backwards compatible and behaves as if it is on an IPv4 only network.

| Network | Dual-Stack | IPv6 | IPv4 | TLS Dual-Stack | TLS IPv6 | TLS IPv4
| -- | -- | -- | -- | -- | -- | --
| IPv4 (Shadow) | IPv4 | Not Possible | Yes | IPv4 | Not Possible | Yes
| IPv6 disabled, Dual-stack+NAT64 | IPv4 | Not Possible | Yes | IPv4 | Not Possible | Yes
| Dual-stack+NAT64 (Astral) | IPv6 | Yes | NAT64 | IPv6 | Yes | NAT64
| IPv6+NAT64 (Wildspace) | IPv6 | Yes | NAT64 | IPv6 | Yes | NAT64

PlatformIO configuration for testing. I have build and uploaded a version of the libs with the required configuration change, so that I can test using PlatformIO:

```ini
platform = https://github.com/sgryphon/platform-espressif32.git#sgryphon/add-esp32-arduino-libs
platform_packages = 
  platformio/framework-arduinoespressif32-libs @ https://github.com/sgryphon/esp32-arduino-libs.git#sgryphon/test-fix-ipv6-config
  platformio/framework-arduinoespressif32 @ https://github.com/sgryphon/arduino-esp32.git#sgryphon/dns-ipv6-check-first-if-have-addr
```

Some example outputs. (These are from my own test app, at https://github.com/sgryphon/iot-demo-build/tree/feature/arduino-esp32-v3-update/m5stack/m5unified_wifi_https)

IPv6-only to dual-stack:

```
[  1524][W][STA.cpp:544] disconnect(): STA already disconnected.
[  1629][D][WiFiNetworkService.cpp:94] loop(): [Network] WiFi begin status 254
[  1759][D][STA.cpp:133] _onStaArduinoEvent(): Arduino STA Event: 13 - STA_CONNECTED
[  1766][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: WF STA connected
[  3530][D][STA.cpp:133] _onStaArduinoEvent(): Arduino STA Event: 17 - STA_GOT_IP6
[  3538][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: WF STA IPv6 fe80:0000:0000:0000:0a3a:f2ff:fe65:db28
[  3552][D][WiFiNetworkService.cpp:47] wifiOnEvent(): [Network] IPv6 address type 2
[  4530][D][STA.cpp:133] _onStaArduinoEvent(): Arduino STA Event: 17 - STA_GOT_IP6
[  4538][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: WF STA IPv6 2407:8800:bc61:1300:0a3a:f2ff:fe65:db28
[  4552][I][Core2Logger.cpp:176] success(): [Core2Logger] Success
[  4558][D][WiFiNetworkService.cpp:47] wifiOnEvent(): [Network] IPv6 address type 1
[  4566][D][STA.cpp:133] _onStaArduinoEvent(): Arduino STA Event: 17 - STA_GOT_IP6
[  4573][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: WF STA IPv6 fd7c:e25e:67e8:0000:0a3a:f2ff:fe65:db28
[  4587][I][Core2Logger.cpp:176] success(): [Core2Logger] Success
[  4593][D][WiFiNetworkService.cpp:47] wifiOnEvent(): [Network] IPv6 address type 4
[ 10488][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: Button 1, scenario 0, v0.1.0-156-gaf345f5
[ 10501][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: Global IPv6 2407:8800:bc61:1300:a3a:f2ff:fe65:db28
[ 10515][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: IPv4 0.0.0.0
[ 10523][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: Link-Local IPv6 fe80::a3a:f2ff:fe65:db28%st1
sta: <UP>
      ether 08:3A:F2:65:DB:28
      inet 0.0.0.0 netmask 0.0.0.0 broadcast 255.255.255.255
      gateway 0.0.0.0 dns [ 10537][D][NetworkInterface.cpp:611] dnsIP(): DNS got IPv6: fd7c:e25e:67e8:0000:0000:0000:0000:0001
fd7c:e25e:67e8::1
      inet6 fe80::a3a:f2ff:fe65:db28%st1 type LINK_LOCAL
      inet6 2407:8800:bc61:1300:a3a:f2ff:fe65:db28 type GLOBAL
      inet6 fd7c:e25e:67e8:0:a3a:f2ff:fe65:db28 type UNIQUE_LOCAL
[ 10568][D][NetworkInterface.cpp:611] dnsIP(): DNS got IPv6: fd7c:e25e:67e8:0000:0000:0000:0000:0001
[ 10583][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: DNS0 fd7c:e25e:67e8::1
[ 10592][D][NetworkInterface.cpp:637] dnsIP(): DNS IPv4: 0.0.0.0
[ 10598][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: DNS1 0.0.0.0
[ 10606][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: URL: http://v4v6.ipv6-test.com/api/myip.php
[ 10619][D][HTTPClient.cpp:303] beginInternal(): protocol: http, host: v4v6.ipv6-test.com port: 80 url: /api/myip.php
[ 10630][D][HTTPClient.cpp:598] sendRequest(): request type: 'GET' redirCount: 0

[ 10637][D][NetworkManager.cpp:91] hostByName(): Clearing DNS cache
[ 10655][D][NetworkManager.cpp:106] hostByName(): DNS found first IPv6 2001:41d0:701:1100::29c8
[ 11049][D][HTTPClient.cpp:1170] connect():  connected to v4v6.ipv6-test.com:80
[ 11462][D][HTTPClient.cpp:1321] handleHeaderResponse(): code: 200
[ 11468][D][HTTPClient.cpp:1328] handleHeaderResponse(): Transfer-Encoding: chunked
[ 11476][D][HTTPClient.cpp:642] sendRequest(): sendRequest code=200

[ 11482][D][HTTPClient.cpp:388] disconnect(): still data in buffer (2), clean up.

[ 11490][D][HTTPClient.cpp:393] disconnect(): tcp keep open for reuse
[ 11496][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: response=<2407:8800:bc61:1300:a3a:f2ff:fe65:db28>
[ 11511][I][Core2Logger.cpp:176] success(): [Core2Logger] Success
```

IPv6-only to IPv4-only, also works via DNS64+NAT64, by using a synthentic IPv6 DNS64 address -- effectively making the destination dual-stack. From the point of view of the destination the source address appears to be the IPv4 NAT address (same as for NAT44).

```
[ 15828][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: Button 3, scenario 2, v0.1.0-156-gaf345f5
[ 15841][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: Global IPv6 2407:8800:bc61:1300:a3a:f2ff:fe65:db28
[ 15855][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: IPv4 0.0.0.0
[ 15863][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: Link-Local IPv6 fe80::a3a:f2ff:fe65:db28%st1
sta: <UP>
      ether 08:3A:F2:65:DB:28
      inet 0.0.0.0 netmask 0.0.0.0 broadcast 255.255.255.255
      gateway 0.0.0.0 dns [ 15877][D][NetworkInterface.cpp:611] dnsIP(): DNS got IPv6: fd7c:e25e:67e8:0000:0000:0000:0000:0001
fd7c:e25e:67e8::1
      inet6 fe80::a3a:f2ff:fe65:db28%st1 type LINK_LOCAL
      inet6 2407:8800:bc61:1300:a3a:f2ff:fe65:db28 type GLOBAL
      inet6 fd7c:e25e:67e8:0:a3a:f2ff:fe65:db28 type UNIQUE_LOCAL
[ 15909][D][NetworkInterface.cpp:611] dnsIP(): DNS got IPv6: fd7c:e25e:67e8:0000:0000:0000:0000:0001
[ 15923][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: DNS0 fd7c:e25e:67e8::1
[ 15932][D][NetworkInterface.cpp:637] dnsIP(): DNS IPv4: 0.0.0.0
[ 15938][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: DNS1 0.0.0.0
[ 15946][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: URL: http://v4.ipv6-test.com/api/myip.php
[ 15956][D][HTTPClient.cpp:303] beginInternal(): protocol: http, host: v4.ipv6-test.com port: 80 url: /api/myip.php
[ 15966][D][HTTPClient.cpp:598] sendRequest(): request type: 'GET' redirCount: 0

[ 15980][D][NetworkManager.cpp:106] hostByName(): DNS found first IPv6 64:ff9b::334b:4e67
[ 16304][D][HTTPClient.cpp:1170] connect():  connected to v4.ipv6-test.com:80
[ 16686][D][HTTPClient.cpp:1321] handleHeaderResponse(): code: 200
[ 16692][D][HTTPClient.cpp:1328] handleHeaderResponse(): Transfer-Encoding: chunked
[ 16700][D][HTTPClient.cpp:642] sendRequest(): sendRequest code=200

[ 16706][D][HTTPClient.cpp:388] disconnect(): still data in buffer (2), clean up.

[ 16714][D][HTTPClient.cpp:393] disconnect(): tcp keep open for reuse
[ 16720][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: response=<220.240.255.134>
[ 16759][I][Core2Logger.cpp:176] success(): [Core2Logger] Success
```

IPv6-only via HTTPS/TLS to dual-stack. 

```
[ 24949][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: Button 4, scenario 3, v0.1.0-160-g02b6d90-dev
[ 24962][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: Global IPv6 2407:8800:bc61:1300:a3a:f2ff:fe65:db28
[ 24973][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: IPv4 0.0.0.0
*sta: <UP,Wildspace,CH:1,RSSI:-57,N,WPA2_PSK> (DHCPC,GARP,IP_MOD,V6_REP)
      ether 08:3A:F2:65:DB:28
      inet 0.0.0.0 netmask 0.0.0.0 broadcast 255.255.255.255
      gateway 0.0.0.0 dns fd7c:e25e:67e8::1
      inet6 fe80::a3a:f2ff:fe65:db28%st1 type LINK_LOCAL
      inet6 2407:8800:bc61:1300:a3a:f2ff:fe65:db28 type GLOBAL
      inet6 fd7c:e25e:67e8:0:a3a:f2ff:fe65:db28 type UNIQUE_LOCAL
[ 25035][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: DNS0 fd7c:e25e:67e8::1
[ 25054][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: DNS1 0.0.0.0
[ 25061][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: TLS URL: https://v4v6.ipv6-test.com/api/myip.php
[ 25075][D][HTTPClient.cpp:303] beginInternal(): protocol: https, host: v4v6.ipv6-test.com port: 443 url: /api/myip.php
[ 25086][D][HTTPClient.cpp:598] sendRequest(): request type: 'GET' redirCount: 0
[ 25094][D][NetworkManager.cpp:106] hostByName(): DNS found IPv6 first 2001:41d0:701:1100::29c8
[ 27088][D][HTTPClient.cpp:1170] connect():  connected to v4v6.ipv6-test.com:443
[ 27702][D][HTTPClient.cpp:1321] handleHeaderResponse(): code: 200
[ 27708][D][HTTPClient.cpp:1328] handleHeaderResponse(): Transfer-Encoding: chunked
[ 27716][D][HTTPClient.cpp:642] sendRequest(): sendRequest code=200
[ 27723][D][HTTPClient.cpp:388] disconnect(): still data in buffer (2), clean up.
[ 27730][D][HTTPClient.cpp:393] disconnect(): tcp keep open for reuse
[ 27736][I][Core2Logger.cpp:198] log(): [Core2Logger] CORE2: response=<2407:8800:bc61:1300:a3a:f2ff:fe65:db28>
[ 27751][I][Core2Logger.cpp:176] success(): [Core2Logger] Success
```


## Related links

Now that the TLS change is merged, once the Libs are also rebuilt (the config is already merged), this addresses the issues in https://github.com/espressif/arduino-esp32/issues/9143, and in the IPv6 discussion.

An alternative fix, that implements RFC6724 destination address selection, has been proposed for LWIP; this would make AF_UNSPEC work and remove the need for the work around checking for IPv6 separately: https://github.com/espressif/esp-lwip/pull/66

sgryphon commented:
I need a bit more testing on this branch after the other changes.

sgryphon commented:
I have re-tested everything with the latest changes.

I also tested a dual-stack network but with IPv6 not enabled, i.e. don't call WiFi.enableIPv6(), and it behaves as if it is on an IPv4 only network, so fully backwards compatible (if code can't handle IPv6 at all, e.g. tries to output to a string buffer that is too small, etc).

I also built the libs with the config changes and pushed to my own repo, so I could try a PlatformIO build with everything set up.

The check for if you have a global IPv6 also now uses Arduino functions like getNetifByID() and hasGlobalIPv6(), rather than low level functions, as it reads nicer.

This is ready to merge, but really needs the libs config update to be of much use.

There are still some problems with TLS / HTTPS, which most production systems should be using, but they need the fix in LWIP, and TLS uses the DNS directly at the lower level.

me-no-dev commented:
The contents of this PR do exactly what `preferV6` used to do for `hostByName`. Get the V6 address if available, else return V4.

sgryphon commented:
> The contents of this PR do exactly what `preferV6` used to do for `hostByName`. Get the V6 address if available, else return V4.

Kind of, except preferV6 was a fixed value, whereas checking IPv6 first is dynamic based on whether you have a public (i.e. global, not link scope) IPv6 address.

If the target is dual stack, then it has both IPv4 and IPv6

If you have IPv4 only, you want the IPv4 address; if you only have IPv6 then you want the IPv6 address. 

So you can't have a fixed preference decided by the programmer when they wrote the code, but have to look at the actual addresses you have been given (depending on what is available on the network and run time).

me-no-dev commented:
@sgryphon you must have not read the code before. It was checking exactly if we have a global address and if so, it allows to set if you prefer V6 or V4 address. You just did not see that code after the refactoring (placeholder left with ToDo)

me-no-dev commented:
Please rebase the PR so it fixes the conflicts.

sgryphon commented:
Rebase done.

> placeholder left with ToDo

Yes, I implemented the ToDo. i.e. the first part of the code that checks if you have a global IPv6.

I also added back in the IPv6 check because the code is now using getaddrinfo(), which has different parameters.

The old code to host passed in either V4_V6 or V6_V4 as a preference order to hostbyname (i.e. manual preference order).

getaddrinfo() doesn't do that -- you pass in AF_UNSPEC for any (which should match your address but the LWIP bug hard codes to V4_V6), or you can limit to one address family only, i.e. AF_INET or AF_INET6.

To implement the preference order, based on actual addresses, the code checks if we have a global IPv6 (no longer a TODO), and if so, looks for AF_INET6 first (if no IPv6, then it doesn't).  If AF_INET6 fails, then it falls back to AF_UNSPEC (which currently prefers IPv4).

sgryphon commented:
... But, I finally worked out how to the Libs building with my LWIP fix, https://github.com/espressif/esp-lwip/pull/66.

With the LWIP fix, then AF_UNSPEC returns the correct destination address selection according to RFC 6724, i.e. if you have an IPv6 it prefers IPv6, so the work around is not needed.

i.e. Current Arduino-ESP32 + LWIP fix works, and this PR is not needed.

If possible, I would prefer the LWIP fix get merged instead, as that is where the issue really needs to be fixed, for all of ESP-IDF (not just Arduino). 

The PlatformIO config, with reference to my build of Libs with the LWIP fix is:

```ini
platform = https://github.com/sgryphon/platform-espressif32.git#sgryphon/add-esp32-arduino-libs
platform_packages = 
  platformio/framework-arduinoespressif32-libs @ https://github.com/sgryphon/esp32-arduino-libs.git#sgryphon/test-fix-ipv6-lwip-and-config
  platformio/framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32.git
```

me-no-dev commented:
i've marked it as blocked, but this PR is not really blocked by anything, is it? I do not find it nice to ask DNS twice. Can we make it optional through `preferV6`?

me-no-dev commented:
BTW. let's split this into two PRs. One will detect if public V6 address exists (and reset DNS cache), the other will do the `preferV6`. I can merge the first one right away

sgryphon commented:
I also looked into the TLS (HTTPS / SSL) issues, which I thought were the same as ESP-IDF, because even with the LWIP fix the TLS connections were still failing.  I have found the problem and have a fix up in another branch; I have tested a few scenarios and it is working, but need to finish testing all scenarios before I put up the PR.

As I would encourage systems to always use HTTPS, getting it working is important.

me-no-dev commented:
Will you please split the PR so I can merge the change of IP detection?

sgryphon commented:
Related TLS fix is here: https://github.com/espressif/arduino-esp32/pull/9470

sgryphon commented:
> Will you please split the PR so I can merge the change of IP detection?

I'm not entirely sure what the IP detection would do by itself, and trying to understand what you want.

sgryphon commented:
Okay, I think I found what you want -- although I don't think it will achieve much.

> One will detect if public V6 address exists (and reset DNS cache), the other will do the preferV6

I can split it like that, although I don't think it will do much. i.e. the first change won't fix anything (I'm not sure what resetting DNS cache will achieve).

The rest of the code is nothing about preferences or preferring, but about solving the problem that if you are IPv6-only then you need to get an IPv6 address to connect to, whereas if you are IPv4 only you need to get an IPv4 address. That is not a preference, but a bug fix to be able to connect IPv6 to a dual-stack destination.

(The only time a preference would apply would be dual-stack to dual-stack where either will work; I don't suggest allowing a preference, but just following RFC6724 that prioritises IPv6.)

So 3 things:
1. Detect IPv6 and refresh cache if changed (I don't think this will do anything useful)
2. Fix the IPv6-only to dual-stack bug, by using the detected IPv6 to determine the order we check DNS.
3. Add a preference option for the dual-stack to dual-stack situation (I don't think this is necessary and would just stick with RFC6724 defaults -- note that the RFC does detail an override mechanism, which is supported in my LWIP code)

me-no-dev commented:
> Detect IPv6 and refresh cache if changed (I don't think this will do anything useful)

very useful if you first got your IPv4 address. Please split the PR

sgryphon commented:
I have split the PRs. The new PR has the IP address type check and DNS cache refresh. https://github.com/espressif/arduino-esp32/pull/9476

This one only adds the IPv6 fix as a second commit (i.e. consider it targetting the other PR).

The full diff for this will contain both changes, but you can look at the individual commits for details.

sgryphon commented:
Rebased on latest main, with just the IPv6 DNS fix commit

sgryphon commented on libraries/Network/src/NetworkManager.cpp:
```diff
@@ -86,12 +86,36 @@ int NetworkManager::hostByName(const char* aHostname, IPAddress& aResult)
         log_d("Clearing DNS cache");
     }
 
-    struct addrinfo hints;
```
Changed.
